### DISCUSSION TITLE
Context and Baggage API usage

### DISCUSSION THREAD

**Discussion by arielvalentin** [open-telemetry/opentelemetry-ruby#701] [Q&A]
I am interested in taking in using the Context and Baggage API in our applications however I want to ensure I am doing so in an idiomatic way. 

I can see that the API provides utility methods like [`OpenTelemetry::Context.with_current`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/context.rb#L44) that changes the [`implicit`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/context.md#overview) reference in [`OpenTelemetry::Context.current`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/context.rb#L28). 

The `OpenTelemetry::Baggage::Manager` also generates a new instance of the `OpenTelemetry::Context` that has a newly populated baggage entry, defaulting to using the `implicit` context. e.g. [`OpenTelemetry::Baggage::Manager#build`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/api/lib/opentelemetry/baggage/manager.rb#L24)

I also see examples in auto-instrumented libraries that set the `implicit` scope using [`OpenTelemetry::Context.with_value`](https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack.rb#L44), but haven't seen any used for `Baggage` just yet. 

Would this be the idiomatic way to do so?
```
updated_context = OpenTelemetry.baggage.build do |builder|
  builder.set_value(key_a: "foo")
  builder.set_value(key_b: "bar")
end

Context.with_current(updated_context) do
   # application code goes here.... 
end
```

**Comment by mwear** [comment]
@arielvalentin your example is idiomatic given the current API. The Baggage::Manager helps to manipulate baggage in a context, and then the Context API is used to activate / deactivate a given context instance.

Also, given that you are starting to look at using baggage, you might have some opinions on https://github.com/open-telemetry/opentelemetry-ruby/discussions/703.

**Reply by arielvalentin** [reply]
I think that the parts here that stand out to me are that the `baggage` convenience method is not referring to a concept or role named `Baggage` but rather the `BaggageManager`, which appears to also be a global variable whose state changes as a result of also changing the implicit `Context.current` value. 

I am not certain if this will make things better, but perhaps if the APIs were symmetric it may reduce what I perceive as friction:
```ruby
OpenTelemetry::Context.with_baggage( key_a: "foo". key_b: "bar" ) do |ctx|
    # your application code goes here
end
```

or 
```ruby
OpenTelemetry.baggage.with_baggage( key_a: "foo". key_b: "bar" ) do |ctx|
    # your application code goes here
end
```
The latter seems off to me because again we do not really have a notion of a `Baggage`, it is a special key in the `Context` so I lean towards the former, but I think that we may be constrained here by the official `API`.

**Comment by ahayworth** [comment]
I almost feel like it would look more idiomatic if it read as:

```ruby
updated_context = OpenTelemetry.baggage.merge do |context|
  context.set_value(key_a: "foo")
  context.set_value(key_b: "bar")
end
```

...primarily since `merge` with a block is something that at least vaguely mimics the `Enumerable` module (granted: this `merge` is not what `Enumerable#merge` really does... but hopefully my point makes it across regardless - maybe `s/merge/update/`, instead).
